File: data-structure/c-cpp/c/Problem/Problem3-1/src/ArrayList.c
#include "ArrayList.h"
#include <stdio.h>

void init(List* list) {
    list -> numOfData = 0;
    list -> curPosition = -1;
}

void insert(List* list, Data data) {
    if (list -> numOfData == LIST_LEN) {
        printf("list is full!!");
    }

    list -> array[list ->numOfData++] = data;
}

int first(List* list, Data* data) {
    if (list -> numOfData == 0) {
        printf("list is empty");
        return FALSE;
    }
    list -> curPosition = 0;
    *data = list -> array[list -> curPosition];
    return TRUE;
}

int next(List* list,Data* data) {
    if (list -> curPosition == list -> numOfData -1) {
        return FALSE;
    }
    list -> curPosition++;
    *data = list -> array[list -> curPosition];
    return TRUE;
}
Data remove_(List* list) {
    Data r_data = list -> array[list -> curPosition];
    if (list -> numOfData > 0) {
        for (int i = list -> curPosition + 1;i<list -> numOfData;i++) {
            list -> array[i-1] = list -> array[i];
        }
    }
    list -> numOfData--;
    list -> curPosition--;
    return r_data;
}
int count(List* list) {
    return list -> numOfData;
}

File: data-structure/java/src/stack/array/ArrayList.java
package stack.array;


import exception.NoElementException;

public class ArrayList <T>{
    private T[] arr;
    private int size;
    private int cur;
    private int capacity;

    ArrayList(){
        this.capacity = 100;
        this.arr = (T[])new Object[this.capacity];
        this.cur = -1;
        this.size = 0;
    }
    public void insert(T data){
        this.arr[this.size++] = data;
    }
    public T first(){
        if(this.size == 0) throw new NoElementException();
        return this.arr[++this.cur];
    }
    public T next(){
        if(this.cur + 1 >= this.size) throw new NoElementException();
        return this.arr[++this.cur];
    }
    public T remove(){
        T rData = this.arr[this.cur];
        for(int i = cur;i<size-1;i++){
            arr[i] = arr[i+1];
        }
        this.cur--;
        this.size--;
        return rData;
    }
    public int count(){
        return this.size;
    }
}
